hyperhtml.js

```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
-      name: attributes[i].name,
-      value: attributes[i].value
-    });
-    return out;
-  }
-
-  // accordingly with the kind of child
-  // it puts its content into a parent node
-  function populateNode(parent, child) {
-    switch (child.nodeType) {
-      case ELEMENT_NODE:
-        var childNodes = parent.childNodes;
-        if (0 < childNodes.length) {
-          if (childNodes[0] === child) {
-            removeNodeList(childNodes, 1);
-            break;
-          }
-        }
-        resetAndPopulate(parent, child);
+  }
+
+  // specify the content to update
+  function setContent(type, target, removeAttributes, childNodes) {
+    var update;
+    switch (type) {
+      case 'any':
+        update = setAnyContent(target);
         break;
-      case DOCUMENT_FRAGMENT_NODE:
-        if (-1 < indexOfDifferences(parent.childNodes, child.childNodes)) {
-          resetAndPopulate(parent, child);
-        }
+      case 'attr':
+        update = setAttribute(target, removeAttributes);
         break;
-      case TEXT_NODE:
-        parent.textContent = child.textContent;
+      case 'text':
+        update = setTextContent(target);
+        break;
+      case 'virtual':
+        update = setVirtualContent(target, childNodes);
         break;
     }
+    return update;
   }
 
-  // it does exactly what it says
-  function removeNodeList(list, startIndex) {
-    var length = list.length, child;
-    while (startIndex < length--) {
-      child = list[length];
-      child.parentNode.removeChild(child);
+  // used for common path creation.
+  function Path(type, node) {
+    return {type: type, path: createPath(node)};
+  }
+
+  // ---------------------------------------------
+  // Hybrid Shims
+  // ---------------------------------------------
+
+  // WeakMap with partial EXPANDO fallback
+  var $WeakMap = typeof WeakMap === typeof $WeakMap ?
+      function () {
+        return {
+          get: function (obj) { return obj[EXPANDO]; },
+          set: function (obj, value) {
+            Object.defineProperty(obj, EXPANDO, {
+              configurable: true,
+              value: value
+            });
+          }
+        };
+      } :
+      WeakMap;
+
+  // Map with partial double Array fallback
+  var $Map = typeof Map === typeof $Map ?
+      function () {
+        var k = [], v = [];
+        return {
+          get: function (obj) {
+            return v[k.indexOf(obj)];
+          },
+          // being used with unique template literals
+          // there is never a case when a value is overwritten
+          // no need to check upfront for the indexOf
+          set: function (obj, value) {
+            v[k.push(obj) - 1] = value;
+          }
+        };
+      } :
+      Map;
+
+  // TODO: which browser needs these partial polyfills here?
+  var isArray = Array.isArray || function (a) { return a instanceof Array; };
+  var trim = EXPANDO.trim || function () { return this.replace(/^\s+|\s+$/g, ''); };
```

Comment:
EXPANDO can only be a string and it will always be a string because it's used as ... `expendo` property.

I am trying to keep an eye on the code size too :wink: